File: ClassEval_T/cpp/solution/NumericEntityUnescaper.cpp

#include <cctype>
#include <sstream>

#include <string>

class NumericEntityUnescaper {
public:
    NumericEntityUnescaper() = default;
    std::string replace(const std::string& input);
    static bool is_hex_char(char c);
};

std::string NumericEntityUnescaper::replace(const std::string& input) {
    std::string out;
    size_t pos = 0;
    size_t length = input.size();
    if (input.empty())
    {
        return "";
    }
    while (pos < length - 2) {
        if (input[pos] == '&' && input[pos + 1] == '#') {
            size_t start = pos + 2;
            bool is_hex = false;

            if (start < length && (input[start] == 'x' || input[start] == 'X')) {
                ++start;
                is_hex = true;
            }

            if (start == length) {
                return out; 
            }

            size_t end = start;
            while (end < length && is_hex_char(input[end])) {
                ++end;
            }

            if (end < length && input[end] == ';') {
                try {
                    int entity_value;
                    std::string number_str = input.substr(start, end - start);
                    std::stringstream ss;
                    if (is_hex) {
                        ss << std::hex << number_str;
                    }
                    else {
                        ss << number_str;
                    }
                    ss >> entity_value;

                    if (ss.fail()) {
                        return out; 
                    }

                    out.push_back(static_cast<char>(entity_value));
                    pos = end + 1;
                    continue;
                }
                catch (const std::exception&) {
                    return out; 
                }
            }
        }

        out.push_back(input[pos]);
        ++pos;
    }

    return out;
}

bool NumericEntityUnescaper::is_hex_char(char c) {
    return std::isdigit(c) || (std::tolower(c) >= 'a' && std::tolower(c) <= 'f');
}

File: ClassEval_T/java/solutuon/NumericEntityUnescaper.java
package org.example;

public class NumericEntityUnescaper {

    public String replace(String string) {
        StringBuilder out = new StringBuilder();
        int pos = 0;
        int length = string.length();

        while (pos < length - 2) {
            if (string.charAt(pos) == '&' && string.charAt(pos + 1) == '#') {
                int start = pos + 2;
                boolean isHex = false;
                char firstChar = string.charAt(start);

                if (firstChar == 'x' || firstChar == 'X') {
                    start++;
                    isHex = true;
                }

                if (start == length) {
                    return out.toString();
                }

                int end = start;
                while (end < length && isHexChar(string.charAt(end))) {
                    end++;
                }

                if (end < length && string.charAt(end) == ';') {
                    try {
                        int entityValue = Integer.parseInt(string.substring(start, end), isHex ? 16 : 10);
                        out.append((char) entityValue);
                        pos = end + 1;
                        continue;
                    } catch (NumberFormatException e) {
                        return out.toString();
                    }
                }
            }
            out.append(string.charAt(pos));
            pos++;
        }

        return out.toString();
    }

    public static boolean isHexChar(char c) {
        return Character.isDigit(c) || ('a' <= Character.toLowerCase(c) && Character.toLowerCase(c) <= 'f');
    }
}